### ISSUE TITLE
Minification can produce variable names which are invalid for kernels

### ISSUE THREAD

**Issue by RobinKa** [gpujs/gpu.js#519]
After a minifier is run some variables are renamed to names such as `$` which are valid in javascript but not valid in kernels. On compilation of the shader an error and a warning (`Error: WebGL warning: shaderSource: String contains the illegal character 0x24.`) will be raised.
Is it possible to have the js -> kernel transpiler detect invalid names and rename them?

**Comment by robertleeplummerjr** [comment]
It is.  Can you reproduce this issue?

**Comment by RobinKa** [comment]
I encountered this problem in one of my projects (an EM field simulation) so it can be reproduced by going to a specific commit there and building / running it. If necessary I can try making a more minimal reproduction.

Here's the commit that still contains this problem: https://github.com/RobinKa/maxwell-simulation/tree/c35717e3b40229f064ddc294cb62e1a2ee85ebaf (later on I just removed the number of variables so it wouldn't generate any illegal ones).

```
git clone https://github.com/RobinKa/maxwell-simulation.git
git checkout c35717e3b40229f064ddc294cb62e1a2ee85ebaf
```

Then remove the `"homepage"` entry (https://github.com/RobinKa/maxwell-simulation/blob/c35717e3b40229f064ddc294cb62e1a2ee85ebaf/package.json#L4) from package.json so you can run it locally.

```
cd maxwell-simulation
npm install
npm run build
serve -s build
```

Then when browsing localhost:5000 I get the following warning and error

```
Error: WebGL warning: shaderSource: String contains the illegal character 0x24.
Error: Error compiling fragment shader: ERROR: 0:? : '' : syntax error
```

The function with lots of variables is here: https://github.com/RobinKa/maxwell-simulation/blob/c35717e3b40229f064ddc294cb62e1a2ee85ebaf/src/App.tsx#L23-L91 and as I mentioned when reducing the number of variables in it the minifier won't produce any illegal variable names and the error is gone.

**Comment by robertleeplummerjr** [comment]
Looking into this.

**Comment by robertleeplummerjr** [comment]
Fixed in https://github.com/gpujs/gpu.js/releases/tag/2.6.5, https://www.npmjs.com/package/gpu.js/v/2.6.5,
Unit tests: https://github.com/gpujs/gpu.js/blob/dbe3b47f8b509ca60ee3b80c46adf379307bae9b/test/issues/519-sanitize-names.js#L9

**Comment by robertleeplummerjr** [comment]
I also took the liberty to work on the typescript robustness for your project.  I found a weakness with typescript in that when using `.setFunctions()` method of either instance of `GPU` or `Kernel` that unless all functions had the same signature, typescript would throw an error, and I believe it to be similarly related to this issue: https://github.com/microsoft/TypeScript/issues/14107, but not completely sure.
I found the compromise to be to use an `any` when using `.setFunctions`, but to still provide the same level of "type strength" when using `.addFunction`.  Mostly this is a note to my future self, if I ever try and revisit.  I did start to have an issue when using arguments on kernels that were of different types, but as well got that properly sorted by inheriting argument types from the `GPU` constructor.
